docs/config.txt
ARACHNE.CFG / arachne.conf documentation
========================================

The Arachne main configuration file is called
ARACHNE.CFG (DOS) or arachne.conf (POSIX).

Each option consists of a keyword and the associated
value. Comments start with a semicolon and extend to
the end of the line. Section headers are ignored.

Boolean options traditionally use 'Yes' and 'No',
but other common terms such as 'True', 'Enable' and
'On' (and 'False', 'Disable' and 'Off', respectively)
are recognized as well. Case does not matter.

Numeric options generally can use decimal, octal
(leading zero) or hexadecimal ('0x' or '0X' prefix)
formats. Note that not all numeric-looking options
are actually numeric.

String options are not quoted.

Settings List
-------------
The following table lists all currently existing options.
Note: Content may be inaccurate.

| file       | option              | type     | default       | comment
+------------+---------------------+----------+---------------+--------------------
| config.c   | CursorType          | char     | 'H'           | [H]and, [C]ross, [A]rrow
| config.c   | SmallIcons          | bool     | Yes           |
| config.c   | Hotkeys             | bool     | Yes           |
| config.c   | AltSysFont          | int      | 0             | 0, 1, 2
| config.c   | BigFont             | bool     | No            |
| config.c   | ScrollBarSize       | int      | 10            | <= 80
| config.c   | Colors              | 2x long  | 11 0          | ink color, paper colors
| config.c   | MouseColors         | 2x long  | 7  15         | darkmouse, brightmouse
| config.c   | Frames              | bool     | No            |
| config.c   | QADT                | bool     | No            | quick-and-dirty tables
| config.c   | CacheFonts          | bool     | Yes           |
| config.c   | Cache2TEMP          | bool     | No            |
| config.c   | Kill468x60          | bool     | No            |
| config.c   | AltKeys             | bool     | No            | DOS only
| config.c   | AltTab              | bool     | No            | (alias for AltKeys)
| config.c   | MailIsDesktop       | bool     | Yes           |
| config.c   | EditHotlistEntry    | bool     | Yes           |
| config.c   | Autodial            | bool     | No            |
| config.c   | VFAT                | bool     | No            | DOS only
| config.c   | KeepHTT             | bool     | Yes           |
| config.c   | VirtualScreen       | int      | 0             | VIRT_SCR, virtual Y size
| config.c   | SmoothScroll        | bool     | No            | VIRT_SCR
| config.c   | ScrollStep          | int      | 1000          | VIRT_SCR and SmoothScroll
| config.c   | ScreenMode          | char     | 'A'           | VIRT_SCR, [A]uto, [S]mart, [N]ice
| config.c   | ScrollBarStyle      | char     | 'A'           | [A]rachne, [W]indows, [N]extstep, ...
| config.c   | ESC                 | char     | Exit          | [I]gnore, [B]ack, else Exit
| config.c   | Multitasking        | char     | Yes           | [N]o, [S]afe, else Yes
| config.c   | Logo                | int      | 2000          |
| config.c   | MinDiskSpace        | long     | 125           | KB
| config.c   | XMS4allgifs         | int      | 256           | KB
| config.c   | XMS4onegif          | long     | 31.25         | KB
| config.c   | ExpireDynamic       | long     | 3600          |
| config.c   | ExpireStatic        | long     | 86400         |
| config.c   | FontShift           | int      | 0             | -2 .. +1, cga default: -2
| config.c   | ASCIIwidth          | int      | 78            | 40 .. 255
| config.c   | PostScriptWidth     | int      | 210           | if NOPS not defined
| config.c   | PostScriptHeight    | int      | 297           | if NOPS not defined
| config.c   | UseKeyMap           | bool     | No            |
| config.c   | Keymap1             | string   | (none)        |
| config.c   | Keymap2             | string   | (none)        |
| config.c   | ScreenSaver         | long     | (none)        |
| config.c   | RefreshDelay        | int      | 20            |
| config.c   | CSS                 | bool     | Yes           |
| config.c   | HTTPreferer         | bool     | Yes           |
| config.c   | HTTPKeepAlive       | bool     | Yes           |
| config.c   | UseProxy            | bool     | No            |
| config.c   | Cookies             | bool     | Yes           |
| config.c   | HTTPS2HTTP          | bool     | No            |
| config.c   | IGNOREJS            | bool     | No            |
| config.c   | IGNOREBASEHREF      | bool     | No            |
| config.c   | AlwaysUseCFGcolors  | bool     | No            |
+------------+---------------------+----------+---------------+--------------------
| init.c     | GraphicsMode        | string   | "Hi16.J"      |
| init.c     | FontPathSuffix      | string   | "iso-8859-1"  | POSIX only
| init.c     | Toolbar             | string   | depends       | DOS: toolbar.cfg / POSIX: templates/toolbar.cfg
| init.c     | History             | string   | "history.lst" |
| init.c     | CacheIndex          | string   | "cache.idx"   |
| init.c     | CookieFile          | string   | "cookies.lst" |
| init.c     | CachePath           | string   | cachepath     | DOS only
+------------+---------------------+----------+---------------+--------------------------
| guievent.c | Hotlist             | string   | "hotlist.htm"       |
| guievent.c | DialPage            | string   | "file:ppp_init.htm" |
| guievent.c | WarnClear           | bool     | No                  |
| guievent.c | PrintURL            | string   | "file:%s%sprint.ah" | (sharepath,GUIPATH)
| guievent.c | SearchPage          | string   | (none)              | call Piip() if unset
| guievent.c | AutoF3key           | bool     | No                  |
| guievent.c | F3key               | string   | "file:*.htm"        |
| guievent.c | RIAD                | bool     | No                  | Reload Index After Delete
| guievent.c | Editor              | ???      | (none)              | checks if "NUL" or not?!?
| guievent.c | HttpReferer         | bool     | No                  |
| guievent.c | ShiftF1 .. ShiftF12 | string   | (none)              |
| guievent.c | Connection          | string   | (none)              | checks if "pppd" or not
+------------+---------------------+----------+---------------+--------------------------
| popget.c   | MailTop             | bool     | No            |
| popget.c   | MailPath            | string   | "MAIL\\"      |
+------------+---------------------+----------+---------------+--------------------
| htmldraw.c | BMPwidth            | int      | 640           | 640 .. 2048
+------------+---------------------+----------+---------------+--------------------
| urlovrl.c  | CachePath           | string   | cachepath     |
| urlovrl.c  | LoadImages          | bool     | Yes           |
+------------+---------------------+----------+---------------+--------------------
| protocol.c | FTPusername         | string   | (empty)       |
| protocol.c | FTPpassword         | string   | (empty)       |
| protocol.c | FTPserver           | string   | (empty)       |
| protocol.c | FTPpath             | string   | (empty)       |
| protocol.c | DialPage            | string   | "file:ppp_init.htm"
| protocol.c | Hangup              | string   | (empty)       |
| protocol.c | ExitOnHangup        | bool     | No            |
| protocol.c | NNTPserver          | string   | (none)        |
| protocol.c | UseFTPproxy         | bool     | No            |
| protocol.c | NoFTPproxy4all      | string   | (none)        |
| protocol.c | NoFTPproxy          | string   | (none)        |
| protocol.c | POP3server          | string   | (none)        |
| protocol.c | POP3username        | string   | (none)        |
| protocol.c | POP3password        | string   | (none)        |
| protocol.c | KeepOnServer        | bool     | No            |
| protocol.c | POP3log             | bool     | No            |
| protocol.c | AfterPOP3           | string   | "file://inbox.dgi"
| protocol.c | SMTPserver          | string   | (none)        |
| protocol.c | SendHELO            | bool     | Yes           |
| protocol.c | SMTPlog             | bool     | No            |
| protocol.c | AuthSMTPusername    | string   | (none)        |
| protocol.c | AuthSMTPpassword    | string   | (none)        |
| protocol.c | UseAuthSMTP         | bool     | No            |
| protocol.c | eMail               | string   | (empty)       |
| protocol.c | MailPath            | string   | (none)        |
| protocol.c | AfterSMTP           | string   | "file://outbox.dgi"
+------------+---------------------+----------+---------------+--------------------
| html.c     | WrapPre             | bool     | No            |
| html.c     | IgnoreImages        | bool     | No            |
| html.c     | MaxImgWidth         | int      | 2048          | >= 100
| html.c     | BGimages            | bool     | Yes           |
| html.c     | IgnoreActive        | bool     | No            |
| html.c     | FTPpath <<tagarg>>  | set to (TAG_ARACHNE_BONUS, NOCACHE, TARGET) arg
+------------+---------------------+----------+---------------+--------------------
| http.c     | AcceptCharset       | string   | (none)        |
| http.c     | NoProxy             | string   | (empty)       | if !strstr(value, host)
| http.c     | NoProxy4all         | string   | (empty)       | if !strstr(value, host)
| http.c     | FTPproxy            | string   | (empty)       |
| http.c     | HTTPproxy           | string   | (empty)       |
| http.c     | ProxyUsername       | string   | (empty)       |
| http.c     | ProxyPassword       | string   | (empty)       |
| http.c     | UserAgent           | string   | "xChaos_Arachne (DOS/5.%s%s)",VER,beta
| http.c     | CachePath           | string   | cachepath     |
| http.c     | UseByteSec          | bool     | Yes           |
+------------+---------------------+----------+---------------+--------------------
| drawtime.c | ScreenSaverColors   | int      | 0             | 0 or 1
| drawtime.c | ScreenSaverMess     | int      | MAX_CERFS/2   | 1 .. MAX_CERFS (20)
| drawtime.c | ScreenSaverStyle    | char     | 'R'           | 'C', 'R'
+------------+---------------------+----------+---------------+--------------------
| guidraw.c  | MaxRes              | int      | 0             |
| guidraw.c  | HomePage            | string   | "http://arachne.cz/" or "file:%shome.htm"
| guidraw.c  | GraphicsMode        | updated with <arachne.graphics> | POSIX only
+------------+---------------------+----------+---------------+--------------------
| cgiquery.c | Profile <<value>>   | string   | (set)         |
| cgiquery.c | KillSent            | bool     | No            |
| cgiquery.c | MailPath            | string   | "MAIL\\"      |
| cgiquery.c | Organization        | string   | (none)        |
| cgiquery.c | PersonalName        | string   | (empty)       |
| cgiquery.c | eMail               | string   | (empty)       |
| cgiquery.c | TimeZone            | string   | "+0000"       |
| cgiquery.c | CC: <<value>>       | string   | (set)         |
| cgiquery.c | BCC: <<value>>      | string   | (set)         |
| cgiquery.c | UseReplyto          | bool     | No            |
| cgiquery.c | RT <<value>>        | string   | (set)         |
| cgiquery.c | <<name>> <<value>>  | ???      | (set)         | form action=arachne:internal... processing
| cgiquery.c | MailEncoding        | string   | "MIME"        |
| cgiquery.c | UseCID              | bool     | No            |
| cgiquery.c | UseCDescr           | bool     | No            |
| cgiquery.c | MyCharset           | string   | "US-ASCII"    |
| cgiquery.c | MailBodyEncoding    | string   | "7bit"        |
| cgiquery.c | UseSignature        | bool     | No            |
| cgiquery.c | SignatureFile       | string   | (none)        |
| cgiquery.c | SavePasswords       | bool     | Yes           |
+------------+---------------------+----------+---------------+--------------------
| adial.c    | IP_Grab             | string   | "IP address set to"
| adial.c    | Hangup              | string   | (empty)       |
| adial.c    | UseTerminal         | bool     | No            |
| adial.c    | TerminalWindow      | string   | (empty)       | if UseTerminal == 'Y' "NUL" means empty
| adial.c    | Dialer              | string   | (empty)       | if UseTerminal == 'N'
| adial.c    | Connection          | string   | (none)        |
| adial.c    | Port                | int      | 0             |
| adial.c    | Irq                 | string   | "0x3f8"       |
| adial.c    | Base                | string   | "4"           |
| adial.c    | NameServer          | string   | (empty)       |
| adial.c    | AltNameServer       | string   | (empty)       |
| adial.c    | Speed               | string   | "19200"       |
| adial.c    | PPPusername         | string   | (empty)       |
| adial.c    | PPPpassword         | string   | (empty)       |
| adial.c    | InitString          | string   | "ATZ"         |
| adial.c    | PhoneNumber         | string   | (empty)       |
| adial.c    | DialPage            | string   | "file:ppp_init.htm"
+------------+---------------------+----------+---------------+--------------------
| urlstat.c  | Index               | string   | "*.htm"       |
| urlstat.c  | KillSent            | bool     | No            |
| urlstat.c  | MailPath            | string   | "MAIL\\"      |
| urlstat.c  | PersonalName        | string   | (empty)       |
| urlstat.c  | eMail               | string   | (empty)       |
| urlstat.c  | TimeZone            | string   | "+0000"       |
| urlstat.c  | MyCharset           | string   | "US-ASCII"    |
| urlstat.c  | SMTPport            | int      | 25            |
| urlstat.c  | POP3port            | int      | 110           |
| urlstat.c  | Mail2Hist           | string   | Yes           | gated behind EXP
+------------+---------------------+----------+---------------+--------------------
| htmlutil.c | AcceptCharset       | string   | (none)        |
| htmlutil.c | IgnoreRefresh       | bool     | Yes           |
| htmlutil.c | ShortestRefresh     | int      | 2             |
| htmlutil.c | HTMLbgColor         | string   | "#C4C4C4"     |
| htmlutil.c | HTMLtext            | string   | "#000000"     |
| htmlutil.c | HTMLlink            | string   | "#0000C4"     |
| htmlutil.c | BgColor             | string   | "#000000"     |
| htmlutil.c | Text                | string   | "#F0F0F0"     |
| htmlutil.c | Link                | string   | "#00FF00"     |
| htmlutil.c | Background          | string   | (none)        | if not "NUL" then filename
| htmlutil.c | DownloadPath        | string   | "DOWNLOAD\\"  |
| htmlutil.c | << tagarg >>        | ARACHNECFGVARIABLE       |
| htmlutil.c | << tagarg >>        | ARACHNECHECKED           |
+------------+---------------------+----------+---------------+--------------------
| atcpip.c   | Connection          | string   | (none)        |
| atcpip.c   | DialPage            | string   | "file:ppp_init.htm"
| atcpip.c   | TCPconfig           | string   | (none)        |
| atcpip.c   | IP_Address          | string   | (none)        |
| atcpip.c   | Gateway             | string   | (none)        |
| atcpip.c   | AltGateway          | string   | (none)        |
| atcpip.c   | Netmask             | string   | (none)        |
| atcpip.c   | NameServer          | string   | (none)        |
| atcpip.c   | AltNameServer       | string   | (none)        |
+------------+---------------------+----------+---------------+--------------------
| linglue.c  | SVGAlib_MouseScale  | int      | 60            |
+------------+---------------------+----------+---------------+--------------------
| main.c     | SearchEngine        | string   | "http://www.google.com/search?q="
+------------+---------------------+----------+---------------+--------------------
| onmouse.c  | RMGB                | bool     | Yes           |
+------------+---------------------+----------+---------------+--------------------
| misc.c     | Hotlist             | string   | "hotlist.htm" |
| misc.c     | JPEGargs            | string   | "-bmp "       |
| misc.c     | << strupr(cfgkw) >> | MIME.CFG $kKEYWORD       |
| misc.c     | Expiredynamic       | (ignored)                |
| misc.c     | FILEargs            | (ignored)                |
| misc.c     | CacheDirList        | MIME.CFG $F -> file browser arguments
| misc.c     | MailPath            | MIME.CFG $M -> mail path
| misc.c     | DownloadPath        | MIME.CFG $W -> download path
| misc.c     | NameServer          | MIME.CFG $N -> first nameserver
| misc.c     | Gateway             | MIME.CFG $G -> first gateway
| misc.c     | Editor              | MIME.CFG $! -> (undocumented)
| misc.c     | Netmask             | MIME.CFG $X -> netmask
| misc.c     | CachePath           | MIME.CFG $A -> cache (archive) path
| misc.c     | Hotlist             | MIME.CFG $B -> bookmarks
+------------+---------------------+----------+---------------+--------------------
| ftp.c      | FakeFTPeMail        | string   | (eMail)       |
| ftp.c      | eMail               | string   | "@"           | if FakeFTPeMail missing
| ftp.c      | UseBinaryFTP        | bool     | No            |
+------------+---------------------+----------+---------------+--------------------
| guitick.c  | Profile             | string   | (empty)       \
| guitick.c  | PersonalName        | string   | (empty)        \
| guitick.c  | eMail               | string   | (empty)         | only shown in
| guitick.c  | AcceptCharset       | string   | (empty)         | user interface
| guitick.c  | Cookies             | string   | (empty)        /
| guitick.c  | KeepOnServer        | string   | (empty)       /
+------------+---------------------+----------+---------------+--------------------


lopif/changes.txt
	LOPIF Graphics Library - Revision History
	-----------------------------------------

990610 : Michael Polak (Arachne Labs) - Original code, released under
         the GNU General Public License.

040612 : Michal H. Tyc - Update of 'zbitbyte.asm', correcting
         the code segment assigned by the linker, performing some
         optimizations and fully commenting the source code.

041017 : Joe da Silva - Revision of 'vesa.h', 'x_mod.c', 'xh_putpx.asm',
         'xh_getpx.asm', 'wrt_vide.asm', 'x_putpix.asm', 'x_getpix.asm',
         'rea_w256.asm', 'x_bar256.asm', 'xh_wrchr.asm', 'xh_write.asm',
         'xh_barx.asm', 'xh_read.asm' and 'wrt_chr.asm', such that the
         VESA bytes/line value is used when available. This corrects a
         compatibility problem with the Matrox Millenium II AGP video
         card in 800x600 modes (and potentially other graphics cards in
         certain SVGA modes). Minor improvements in error handling were
         also made in a couple of places. Minor improvements were also
         made to the 'vv.cc' test application code.


config.c
#include "arachne.h"
#include "internet.h"

/*

void defaultGUIstyle(void)
{

  if(x_maxx()<640)
   arachne.GUIstyle = STYLE_SMALL2;
  else
  if(x_maxx()<800)
   arachne.GUIstyle = STYLE_SMALL1;

}

*/
#ifndef POSIX
void meminit(char arg)
{
 switch(arg)
 {
  case 1: //EMS
  SetXmsEms(2);
  break;
  case 2: //disk
  DisableXMS=1;
  break;
  default://like previous versions
  SetXmsEms(1);
 }
}
#endif

/**
 * save pick file (runtime data)
 */
void savepick()
{
	char *fname;
	int fd;

	fname = newstr("%s%s", userpath, "arachne.pck");
	fd = a_open(fname, O_WRONLY | O_CREAT | O_TRUNC, S_IREAD | S_IWRITE);
	freestr(fname);

	if (fd >= 0) {
		(void)write(fd, &arachne, sizeof(struct ArachnePick));
		(void)write(fd, p->htmlframe, MAXFRAMES * (1 + sizeof(struct HTMLframe)));
		(void)write(fd, AUTHENTICATION, sizeof(struct AUTH_STRUCT));
		a_close(fd);
	}
}

//in case newvalue!=NULL
//keystring must be allocated at the length keystring+newvalue together!

static char *configvariable(struct ib_editor *fajl,char *keystring,char *newvalue)
{
 int l;
 char *line,*ptr;

 l=strlen(keystring);
 fajl->y=0;
 while(fajl->y<fajl->lines)
 {
reaquire://!!glennmcc: Mar 08, 2008 -- case sensitive for entity.cfg
  line=ie_getline(fajl,fajl->y);
  if(line && strlen(line)>l && !strncmpi(line,keystring,l))//original line
  {
//!!glennmcc: Mar 08, 2008 -- case sensitive for entity.cfg
   char *filename=fajl->filename;
   if(strstr(filename,"entity.cfg") && strncmp(line,keystring,l))
     {
      fajl->y++;
      goto reaquire;
     }
//!!glennmcc: end
   if(newvalue)
   {
    newval:
    line=farmalloc(IE_MAXLEN);
    if(line)
    {
     strcpy(line,keystring);
     strcat(line," ");
     strcat(line,newvalue);
     ie_putline(fajl,fajl->y,line);
     farfree(line);
    }
    return NULL;
   }
   else
   {
    line+=l;                 //ted ukazuju na prvni mezeru
      // tr.: now I point at first space
    while(*line==' ')line++; //uriznu mezeru
      // tr.: I cut off space
    ptr=strchr(line,';');
    if(ptr)                  //je za hodnotou komentar ?
      // tr.: is after value a comment?
    {
     while(*(ptr-1)==' ')ptr--;//uriznu mezeru pred komentarem
      // tr.: I cut off space before comment
     *ptr='\0';
    }
    return line;             //vracim pointer na hodnotu promenne.
      // tr.: I reset pointer to the value of the variable
   }
  }
  fajl->y++;
 }
 if(newvalue && *newvalue && fajl->y<fajl->maxlines)
 {
  ie_insline(fajl,fajl->y,"");
  goto newval;
 }

 return NULL;
}

/**
 * read boolean option with default
 *   returns 1 for "Yes" / "True"  / "Enable"  / "On"
 *   returns 0 for "No"  / "False" / "Disable" / "Off"
 */
int config_get_bool(char *key, int defval)
{
	char *ptr = configvariable(&ARACHNEcfg, key, NULL);
	if (ptr) {
		switch (toupper(ptr[0])) {
		case 'Y': /* Yes    */
		case 'T': /* True   */
		case 'E': /* Enable */
			return 1;

		case 'N': /* No      */
		case 'F': /* False   */
		case 'D': /* Disable */
			return 0;
		}

		/* On / Off */
		if (!strcmpi(ptr, "on"))
			return 1;
		if (!strcmpi(ptr, "off"))
			return 0;
	}

	/* return default */
	return defval ? 1 : 0;
}

/**
 * read int option with default
 *   supports decimal, octal and hexadecimal
 */
int config_get_int(char *key, int defval)
{
	long val = config_get_long(key, defval);
	return (int)val;
}

/**
 * read long option with default
 *   supports decimal, octal and hexadecimal
 */
long config_get_long(char *key, long defval)
{
	char *ptr = configvariable(&ARACHNEcfg, key, NULL);
	if (ptr) {
		char *eptr;
		long val = strtol(ptr, &eptr, 0);
		if (val || ptr != eptr) {
			return val;
		}
	}
	return defval;
}

/**
 * read string option with default
 */
char *config_get_str(char *key, char *defstr)
{
	char *ptr = configvariable(&ARACHNEcfg, key, NULL);
	if (ptr) {
		return ptr;
	}
	return defstr;
}

/**
 * update string option with new value
 */
void config_set_str(char *key, char *newval)
{
	configvariable(&ARACHNEcfg, key, newval);
}

/**
 * get entity value or NULL
 */
char *config_get_entity(char *entity)
{
	char *ptr = configvariable(&ENTITYcfg, entity, NULL);
	return ptr;
}

/**
 * get toolbar string or NULL
 */
char *config_get_toolbar(char *name)
{
	char *ptr = configvariable(&TOOLBARcfg, name, NULL);
	return ptr;
}

#ifndef WWWMAN
#ifndef POSIX

//. Roura : name of a batch file.  Allow us to use external executable.  Like Pipe in UNIX
void rouraname(char *fname)
{
 tempinit(fname);
 strcat(fname,"$roura$.bat");
}

// create a batch file executing cmd , uses name created by rouraname( ..)
int willexecute(char *cmd) //vykonat nejakou hloupost:
  // tr.: to do foolish things
{
 int f;
 char bat[80];

 rouraname(bat);
 f=a_fast_open(bat,O_TEXT|O_WRONLY|O_CREAT|O_TRUNC,S_IREAD|S_IWRITE);
 if(f>=0)
 {
  write(f,cmd,strlen(cmd));
  a_close(f);
  if(tempdir[0])
   return EXIT_FAST_EXECUTE; // %ARACHNETEMP%\$roura$.bat
  else
   return EXIT_EXECUTE;      // .\$roura$.bat
 }
 else
 {
  return 0;
 }
}
#endif

extern struct ib_editor ARACHNEcfg;// main configuration

void configure_user_interface(void)
{
 char *value;
 long numval;

 /* Hand cursor */
 static const unsigned short cur[32] = {
  0x9FFF, 0x0FFF, 0x07FF, 0x83FF, 0xC1FF, 0xE0FF, 0xF067, 0xF003,
  0xF001, 0xF000, 0xF800, 0xF800, 0xF800, 0xFC00, 0xFC00, 0xFC00,
  0x0000, 0x6000, 0x7000, 0x3800, 0x1C00, 0x0E00, 0x0700, 0x0018,
  0x07EC, 0x07EE, 0x001E, 0x03EE, 0x03EE, 0x001E, 0x00EC, 0x0002
 };

 /* Cross cursor */
 static const unsigned short cur1[32] = {
  0xFC7F, 0xFC7F, 0xFC7F, 0xFC7F, 0xFC7F, 0xFC7F, 0x0001, 0x0101,
  0x0001, 0xFC7F, 0xFC7F, 0xFC7F, 0xFC7F, 0xFC7F, 0xFC7F, 0xFFFF,
  0x0100, 0x0100, 0x0100, 0x0100, 0x0100, 0x0100, 0x0100, 0xFEFE,
  0x0100, 0x0100, 0x0100, 0x0100, 0x0100, 0x0100, 0x0100, 0x0000
 };

 /* Arrow cursor */
 static const unsigned short cur2[32] = {
  0x1FFF, 0x0FFF, 0x07FF, 0x03FF, 0x01FF, 0x00FF, 0x007F, 0x003F,
  0x001F, 0x01FF, 0x10FF, 0x30FF, 0x787F, 0xF87F, 0xFC3F, 0xFFFF,
  0x4000, 0x6000, 0x7000, 0x7800, 0x7C00, 0x7E00, 0x7F00, 0x7F80,
  0x7C00, 0x6C00, 0x4600, 0x0600, 0x0300, 0x0300, 0x0180, 0x0000
 };

 value = config_get_str("CursorType", "Hand");
 switch (toupper(*value)) {
 case 'A':
  x_defcurs(cur2, &cur2[16], 15); // 'Arrow' cursor
  break;
 case 'C':
  x_defcurs(cur1, &cur1[16], 15); // 'Cross' cursor
  break;
 case 'H':
 default:
  x_defcurs(cur, &cur[16], 15); // 'Hand' cursor
  break;
 }

 user_interface.iconsoff = !config_get_bool("SmallIcons", 1);
 user_interface.hotkeys = config_get_bool("Hotkeys", 1);

 numval = config_get_long("AltSysFont", 0);
 if (numval < 0) numval = 0;
 if (numval > 2) numval = 2; /* prevent lock up in 640x480 */
 if (numval > 0) {
  user_interface.bigfnum  = numval;
  user_interface.bigfont  = 1;
  user_interface.bigstyle = FIXED;
 }

 if (config_get_bool("BigFont", 0)) {
  user_interface.bigfont  = 1;
  user_interface.bigfnum  = 5;
  user_interface.bigstyle = BOLD|FIXED;
 }

 user_interface.scrollbarsize = config_get_long("ScrollBarSize", 10);
 if(user_interface.scrollbarsize > 80)
  user_interface.scrollbarsize=80;

 value = config_get_str("Colors", "11 0"); /* green-on-black */
 if(value)
 {
  char *newvalue;

  user_interface.ink = (int)strtol (value, &newvalue, 10);
  if (newvalue != value) {
    value = newvalue;
    user_interface.paper = (int)strtol (value, &newvalue, 10);
    if (newvalue == value)
       user_interface.paper=0; //black
  }
  else user_interface.ink=11; //green
 }

 value = config_get_str("MouseColors", "7 15"); /* grey or white */
 if(value)
 {
  char *newvalue;

  user_interface.darkmouse = (int)strtol (value, &newvalue, 10);
  if (newvalue != value) {
    value = newvalue;
    user_interface.brightmouse = (int)strtol (value, &newvalue, 10);
    if (newvalue == value)
       user_interface.brightmouse=15; //black
  }
  else user_interface.darkmouse=7; //white
 }

 user_interface.frames = config_get_bool("Frames", 0);
 user_interface.quickanddirty = config_get_bool("QADT", 0);
 user_interface.cachefonts = config_get_bool("CacheFonts", 1);
 user_interface.cache2temp = config_get_bool("Cache2TEMP", 0);
 user_interface.killadds = config_get_bool("Kill468x60", 0);

#ifndef POSIX
 if (config_get_bool("AltKeys", 0) ||
     config_get_bool("AltTab", 0)) {
  InstalAltTab();
  atexit(ReleaseAltTab);
 }

 InstalPrtScr();
 atexit(ReleasePrtScr);
#endif

 user_interface.mailisdesktop = config_get_bool("MailIsDesktop", 1);
 user_interface.edithotlistentry = config_get_bool("EditHotlistEntry", 1);
 user_interface.autodial = config_get_bool("Autodial", 0);

#if defined(MSDOS) && !defined(XTVERSION)
 user_interface.vfat = config_get_bool("VFAT", 0);
#endif

 user_interface.keephtt = config_get_bool("KeepHTT", 1);
 user_interface.virtualysize = 0;
 user_interface.smooth = 0;

#ifdef VIRT_SCR
 user_interface.virtualysize = config_get_int("VirtualScreen", 0);
 user_interface.smooth = config_get_bool("SmoothScroll", 0);
 if (user_interface.smooth) {
  user_interface.step = config_get_int("ScrollStep", 1000);
 }

 /* [A]uto, [S]mart, [N]ice */
 value = config_get_str("ScreenMode", "Auto");
 switch (toupper(*value)) {
 case 'S':
 case 'N':
  user_interface.screenmode = toupper(*value);
  break;
 case 'A':
 default:
  user_interface.screenmode = 0;
  break;
 }
#endif // VIRT_SCR

 value = config_get_str("ScrollBarStyle", "Arachne");
 switch (toupper(*value)) {
 case 'A':
  user_interface.scrollbarstyle = 0; /* [A]rachne */
  break;
 case 'C':
  user_interface.scrollbarstyle = 'C'; /* [C]larence */
  user_interface.scrollbarsize = 0;    /* no scrollbars */
  break;
 default:
  /* [W]indoze, [N]extstep, [X]perimental, ... */
  user_interface.scrollbarstyle = toupper(*value);
  break;
 }

 value = config_get_str("ESC", "X");
 switch (toupper(*value)) {
 case 'I':
  user_interface.esc = ESC_IGNORE;
  break;
 case 'B':
  user_interface.esc = ESC_BACK;
  break;
 default:
  user_interface.esc = ESC_EXIT;
  break;
 }

 value = config_get_str("Multitasking", "X");
 if (toupper(*value) == 'S') {
  user_interface.multitasking = MULTI_SAFE;
 } else {
  int val = config_get_bool("Multitasking", 1);
  user_interface.multitasking = val ? MULTI_YES : MULTI_NO;
 }

 user_interface.logoiddle = config_get_int("Logo", 2000);
 user_interface.mindiskspace = config_get_long("MinDiskSpace", 128) * 1024;
 user_interface.xms4allgifs = config_get_int("XMS4allgifs", 256);
 user_interface.xms4onegif = config_get_long("XMS4onegif", 32) * 1024;
 user_interface.expire_dynamic = config_get_long("ExpireDynamic", 3600);
 user_interface.expire_static = config_get_long("ExpireStatic", 86400L);

 if (cgamode) {
  user_interface.fontshift = -2;
 } else {
  user_interface.fontshift = config_get_long("FontShift", 0);
  if (user_interface.fontshift < -2) user_interface.fontshift = -2;
  if (user_interface.fontshift >  1) user_interface.fontshift =  1;
 }

 user_interface.printerwidth = config_get_long("ASCIIwidth", 78);
 if (user_interface.printerwidth <  40) user_interface.printerwidth = 40;
 if (user_interface.printerwidth > 254) user_interface.printerwidth = 255;

#ifndef NOPS
 user_interface.postscript_x = config_get_int("PostScriptWidth",  210); /* A4 */
 if (user_interface.postscript_x <  100) user_interface.postscript_x = 100;
 if (user_interface.postscript_x > 1000) user_interface.postscript_x = 1000;

 user_interface.postscript_y = config_get_int("PostScriptHeight", 297); /* A4 */
 if (user_interface.postscript_y <  100) user_interface.postscript_y = 100;
 if (user_interface.postscript_y > 1000) user_interface.postscript_y = 1000;
#endif

 user_interface.keymap = NULL;
 if (config_get_bool("UseKeyMap", 0)) {
  value=configvariable(&ARACHNEcfg,"Keymap1",NULL);
  if(value && *value)
  {
   user_interface.keymap=farmalloc(96);
   if(user_interface.keymap)
   {
    strncpy((char *)user_interface.keymap,value,48);
    value=configvariable(&ARACHNEcfg,"Keymap2",NULL);
    if(value && *value)
     strncpy((char *)&user_interface.keymap[48],value,48);
   }
  }
 }

 ScreenSaver = config_get_long("ScreenSaver", 0);
 user_interface.refresh = config_get_int("RefreshDelay", 20);

 user_interface.css = config_get_bool("CSS", 1);

 // HTTP parameters --------------------------------------------------

 http_parameters.referer = config_get_bool("HTTPreferer", 1);
 http_parameters.keepalive = config_get_bool("HTTPKeepAlive", 1);
 http_parameters.useproxy = config_get_bool("UseProxy", 0);
 http_parameters.acceptcookies = config_get_bool("Cookies", 1);
 http_parameters.https2http = config_get_bool("HTTPS2HTTP", 0);
 user_interface.ignorejs = config_get_bool("IGNOREJS", 0);
 http_parameters.ignorebasehref = config_get_bool("IGNOREBASEHREF", 0);
 user_interface.alwaysusecfgcolors = config_get_bool("AlwaysUseCFGcolors", 0);
}

#endif
